### ISSUE TITLE
Use Babel runtime

### ISSUE THREAD

**Issue by dozoisch** [react-bootstrap/react-bootstrap#579]
A lot of components have duplicated code in their generated code.

Ex:
![badge](https://cloud.githubusercontent.com/assets/4422516/7341363/181602c4-ec6f-11e4-86c6-d33758ea1101.png)

We could fix that, by adding [babel-runtime][babel] in the build. This way, it requires babel-runtime modules instead of duplicating the output. This is one more dep for us but could help reduce the lib output size.

[babel]:https://babeljs.io/docs/usage/runtime/

**Comment by AlexKVal** [comment]
This warning
> babel-runtime does not support AMD module loaders like RequireJS.

needs diligent research.

I think we can get answer for the question via:
- comparing size of `react-bootstrap.min.js` with and without the `babel-runtime`
- comparing performance with [JSPerf](https://jsperf.com/) for the most used components with and without the runtime

I think the answer will become obvious with these numbers then.

**Comment by dozoisch** [comment]
My concern is also for the lib generated folder that gets required by commonJS users. So we have to check all the generated dists and see the difference

**Comment by mtscout6** [comment]
We could do babel-runtime for the commonjs build and not do it for amd, it would be worth noting to amd users that they'd be using a sub-optimal approach.

**Comment by mtscout6** [comment]
We could do babel-runtime for the commonjs build and not do it for amd, it would be worth noting to amd users that they'd be using a sub-optimal approach.

**Comment by AlexKVal** [comment]
> they'd be using a sub-optimal approach.

It's not the fact :)
Only numbers -> `size change` for min.js and `JSPerf`ormance can tell it for sure.
Don't you think ?

Or you have been done those checks one day already and you know the answers and just let us find out them too ? :smile:

**Comment by mtscout6** [comment]
I have not run the numbers, and the difference may be negligible. I was thinking about needing to declare those functions at runtime in every module as opposed to declaring them once in babel-runtime. But like I said that runtime difference may be negligible and not a big deal.

**Comment by dozoisch** [comment]
Okay, I've compared the two minified size for the fun of it and we're saving about 4kB

- without runtime 101323 (99kB)
- with 97059 (95kB)

So it's not really a big deal, though I could see that being important for some users since we're a lib they package with their stuff.

**Comment by dozoisch** [comment]
Will close because I really don't have the time to do perf testing and everything. Will reopen if I ever get the time.

**Comment by taion** [comment]
I'm re-opening this. I think it's a really good idea to switch to `babel-runtime`, and the docs even explicitly say that `babel-runtime` is a good idea for libraries. I don't see any mentions of performance issues with the runtime, so I expect it not to be a big deal.

And if I'm not mistaken, we're killing the AMD build anyway, so that should be a non-issue.

**Comment by AlexKVal** [comment]
@taion just for ref additional discussion was about it here https://github.com/react-bootstrap/react-bootstrap/pull/751#issuecomment-106557760
and it is interesting a bit, I've got that time another results with sizes than https://github.com/react-bootstrap/react-bootstrap/issues/579#issuecomment-96821142

**Comment by taion** [comment]
Interesting and unexpected. I'd have thought it would have just replaced existing in-line definitions with the imported ones.

**Comment by AlexKVal** [comment]
Maybe I'm doing it wrong.

I'm adding it into `webpack.config` as
`loader: 'babel-loader?optional[]=runtime'` (and of course `&cacheDirectory=${babelCache}`)
and then `npm run build`
Then distr files increase in size and `lib` files stay as is, without changes.

For `lib` files to change it needs to add into `.babelrc`
```js
{
  "optional": [
    "es7.objectRestSpread",
    "runtime"
  ]
}
```
Then `lib` files change to `require('babel-runtime')`, but `distr` files anyway increase size.

If I only add `.babelrc` option and don't change `webpack.config`,
then `distr` files are w/o changes at all. Only `lib` files are changed.

I've tried to add `babel-runtime` into `"externals"` section in `webpack.conf`, but without success :sweat:

**Comment by jquense** [comment]
The runtime does, in my opinion, two different things.

1. polyfill future features, like symbol, Promise, Iterators, etc. in a way that doesn't pollute the Global scope
2. Move the few babel helpers from being inlined in each file to a module that is required instead of inlining.

I tend to think that if we aren't careful with 1. we start to pass the costs of transpiling to our users in the form of lots of extra bytes.  In most cases we are writing es6 code because it is convenient to us a the devs, but it doesn't offer any measurable benefit to a consumer whether the code is es5/es6/es2021. However, if we start using features that need to be polyfilled in the majority of browsers, the whole thing starts to _cost_ the user, because they need to ship a lot more code.

In terms of number 2, we can actually get that benefit without the runtime, with a pretty simple babel plugin and a small change to our babel build script, I was thinking about working on a PR for that to demonstrate.

**Comment by mtscout6** [comment]
If it's a matter of costing our downstream users due to global pollution than I think the decision to use the runtime is a no brainer even if the size increases a bit. I also think that if it's an npm dependency than with npm v3 it may get flattened and downstream users applications would be the ones to see slightly smaller file sizes in the long run. My guess is that since our library is small in comparison to some larger apps we won't see improved file sizes, but I'd venture to guess that there would be some threshold where you'd see a difference.

**Comment by taion** [comment]
Or just run `npm ddp` - end up potentially saving bytes if they're used elsewhere v. using custom polyfills.

I'd be fine with saying "don't use features that have to be polyfilled" as a style issue, but I don't think avoiding accidentally using the polyfills would be a good reason to re-write the Babel runtime plugin, regardless of quantity of code.

**Comment by jquense** [comment]
> if it's a matter of costing our downstream users due to global pollution than I think the decision to use the runtime is a no brainer even if the size increases a bit.

That isn't the concern here I don't think, if we are using things that need to be polyfilled then, we absolutely should use the runtime, vs polyfilling ourselves. The question is more whether we should use things that need to be polyfilled in the first place, of which I am generally of the opinion that we should not (minus the es5 features that ie8 needs).

> avoiding accidentally using the polyfills would be a good reason to re-write the Babel runtime plugin, regardless of quantity of code

It's not rewriting the runtime plugin...maybe it has changed recently but the runtime transformer was a lot less percise when inline helpers, inserting _all_ helpers regardless of whether you used them or not. It also in inadvertently pulled all of CoreJs in regardless of whether you were using it.

**Comment by jquense** [comment]
Just did an initial test with the newer version...helpers are more modular with is great it just imports the specific one it needs, but Babel also seems to also repalce calls that I wouldn't expect like:
- Object.keys,
- Object.defineProperty (may go away if we enable loose mode)

I am not sure why it does that, or how much code that pulls in of CoreJS

**Comment by taion** [comment]
I see the bundles as still being ~10% smaller with runtime than without.

I do see that just changing `.babelrc` is not sufficient, though, which is not ideal. Not sure why this is happening.

**Comment by taion** [comment]
Oh, duh, I need to wipe the cache first.

I see with `babel-runtime`:

```
jjia@jjia-mbp:~/vendor/react-bootstrap (master *=)$ ll dist/*
-rw-r--r--  1 jjia  staff  252684 Jun 15 12:16 dist/react-bootstrap.js
-rw-r--r--  1 jjia  staff  115293 Jun 15 12:16 dist/react-bootstrap.min.js
-rw-r--r--  1 jjia  staff  607544 Jun 15 12:16 dist/react-bootstrap.min.js.map
```

Against master:

```
jjia@jjia-mbp:~/vendor/react-bootstrap (master =)$ ll dist/*
-rw-r--r--  1 jjia  staff  243575 Jun 15 12:18 dist/react-bootstrap.js
-rw-r--r--  1 jjia  staff  118942 Jun 15 12:18 dist/react-bootstrap.min.js
-rw-r--r--  1 jjia  staff  562565 Jun 15 12:18 dist/react-bootstrap.min.js.map
```

**Comment by jquense** [comment]
:+1: glad my info is out of date here :) looks like it is providing a benefit in terms of size.

There is still the more general question whether we use things that need polyfills in all browsers?

**Comment by AlexKVal** [comment]
252,684 - the size WITH the `babel-runtime`
and
233,276 - the size without.

So there is no size benefit :cherries:

**Comment by taion** [comment]
@jquense "All browsers" is a bit tricky here - e.g. "Symbol" doesn't need a polyfill on Chrome, but you need to include the weight for polyfill anyway.

@AlexKVal The minified bundle is smaller, though, which is all we care about.

**Comment by AlexKVal** [comment]
nope :) 113,606 is minified

**Comment by AlexKVal** [comment]
Ups... My bad. Had to clean cache. 
the real size 118,942. You are right. 
The minified version is a bit smaller.
:blush:

**Comment by mtscout6** [comment]
Do we need to add a cache clean into the build or release process?

**Comment by AlexKVal** [comment]
It seems it does need.

**Comment by taion** [comment]
It'd probably be a good idea for the release process (or just don't use the cache for release, to verify correctness). It'd defeat the point of having the cache to introduce it to the build process.

**Comment by taion** [comment]
For funsies, I checked what the size would look like if we added a use of `Symbol`:

```
jjia@jjia-mbp:~/vendor/react-bootstrap (master *=)$ ll dist/*
-rw-r--r--  1 jjia  staff  259501 Jun 15 12:35 dist/react-bootstrap.js
-rw-r--r--  1 jjia  staff  117801 Jun 15 12:36 dist/react-bootstrap.min.js
-rw-r--r--  1 jjia  staff  626374 Jun 15 12:36 dist/react-bootstrap.min.js.map
```

At 117,801B v 115,293B (compare to 118,942B without `babel-runtime`), the weight seems non-trivial.

That said, I'd just add `babel-runtime` for now, and defer discussion regarding polyfilled features.

**Comment by taion** [comment]
Opened https://github.com/react-bootstrap/react-bootstrap/issues/840 regarding the Babel cache issues.

**Comment by AlexKVal** [comment]
Update for posterity:
I had to clean babel-cache to get the right results.

**Comment by AlexKVal** [comment]
Resolved by #841
